ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -117,76 +132,42 @@ delayedProgressReporting before after (Just lspEnv) optProgressStyle = do
     progressState <- newVar NotStarted
     let progressUpdate event = updateStateVar $ Event event
         progressStop   =  updateStateVar StopProgress
-        updateStateVar = modifyVar_ progressState . updateState (lspShakeProgress inProgressState)
-
+        updateStateVar tran = do
+            start <- lspShakeProgressNew inProgressState
+            modifyVar_ progressState $ updateState start tran
         inProgress = updateStateForFile inProgressState
     return ProgressReporting{..}
     where
-        lspShakeProgress InProgressState{..} = do
+        lspShakeProgressNew InProgressState{..} = do
             -- first sleep a bit, so we only show progress messages if it's going to take
```

Comment:
Ah, I would actually really like to do this in `lsp` as well, see https://github.com/haskell/lsp/issues/549